README.md
# generator-packing

一个快速生成 [Packing](https://packingjs.github.io/) 工程的手脚架工具。

## 特点
* 不依赖 host 文件，根据环境自动切换资源路径
* 节约开发服务器，多分支开发部署到同一台服务器不会相互覆盖
* 提供资源包体积分析报告
* 自动生成模版文件

## 使用步骤
1. 全局安装`yo`和`generator-packing`
  ```
  npm install -g yo generator-packing
  ```

1. 在目标目录运行以下命令，该命令会帮助你完成工程初始化和依赖包安装
  ```
  yo packing
  ```

1. 启动开发模式
  ```
  npm run serve
  ```
  如果上面的代码运行不报错的话，就可以在 `http://localhost:8081` 预览网站了

## 命令
* **npm start**: 启动开发模式
* **npm start:dist**: 预览编译后的工程
* **npm build**: 本地编译
* **npm build:dev**: 编译到开发机
* **npm build:beta**: 编译到测试机
* **npm build:prod**: 编译到线上机
* **npm serve**: `npm start` 的别名
* **npm serve:dist**: `npm start:dist` 的别名
* **npm lint**: 代码检查

## 目录结构

```
.
├── /assets/      # 图片字体等资源
│   ├── /images/  # 图片（示例）
│   └── /fonts/   # 字体（示例）
├── /config/
│   ├── /packing.js             # packing配置
│   ├── /webpack.build.babel.js # webpack编译配置
│   ├── /webpack.dll.babel.js   # webpack DllPlugin编译配置
│   └── /webpack.serve.babel.js # webpack开发环境配置
├── /mock/
│   ├── /api              # 异步请求接口模拟数据存放目录
│   │   └── /now.js       # /api/now接口模拟数据（示例）
│   └── /pages            # 初始化网页的模拟数据
│       ├── /__global.js  # 所有页面初始化通用的数据
│       ├── /index.js     # /index页面初始化的数据（示例）
│       └── /about.js     # /about页面初始化的数据（示例）
├── /profiles/
│   ├── /beta.env         # 测试环境profile文件
│   ├── /development.env  # 开发环境profile文件
│   ├── /local.env        # 本地环境profile文件
│   └── /production.env   # 线上环境profile文件
├── /src/                           
│   ├── /common/                    # 公共代码
│   ├── /pages/                     # DllPlugin插件编译配置
│   │   ├── /index/                 # 首页的资源文件（示例）
│   │   │   ├── /entry.js           # entry pointer
│   │   │   └── /entry.settings.js  # 自动生成模版使用的配置文件（可选）
│   │   └── /about/                 # about页的资源文件（示例）
│   │   │   ├── /entry.js           # entry pointer
│   │   │   └── /entry.settings.js  # 自动生成模版使用的配置文件可选）
│   └── /templates                  # 模版文件
│       ├── /layout/                # （pug用的）模版布局文件
│       └── /pages/                 # 用来生成页面的模版文件
├── /util/            # util
├── .babelrc          # babel配置
├── .env              # NODE_ENV环境变量配置，该文件不要提交到git仓库
├── .eslintrc.js      # eslint配置
├── build.sh          # jenkins发布调用的脚本
├── pom.xml           # 前后端关联maven配置
├── postcss.config.js # postcss配置
└── README.md
```

## packing.js 配置说明
```js
{
  /**
   * 工程使用的编译平台
   * 目前支持
   * - portal
   * - qdr
   * @type {string}
   */
  ci: 'portal',

  /**
   * 本地访问的域名
   * 如果需要使用 `qunar.com` 的 cookie，需要改成类似 `my.qunar.com` 这种
   * @type {string}
   */
  localhost: 'localhost',

  /**
   * webserver 端口
   */
  port: {
    /**
     * 开发环境 webserver 端口
     * @type {number}
     */
    dev: 8081,

    /**
     * 预览编译结果时 webserver 端口
     * @type {number}
     */
    dist: 8080
  },

  /**
   * 文件路径配置
   * 所有目录都使用相对于项目根目录的相对目录格式
   */
  path: {
    /** 源文件相关路径 */
    src: {
      /**
       * 源文件根目录
       * @type {string}
       */
      root: 'src',

      /**
       * 模版文件路径
       * 相对于 `src.root` 的相对地址
       * 若不区分布局文件和网页文件，请直接传入字符串
       * @type {(string|object)}
       */
      templates: {
        layout: 'templates/layout',
        pages: 'templates/pages'
      }
    },

    /** 编译输出文件相关路径 */
    dist: {
      /**
       * webpack 编译产物输出目录
       * 即 `webpack.config.output.path` 参数
       * portal dev 发布时要求输出到 `dev` 目录
       * qdr dev 发布时要求输出到 `prd` 目录
       * @type {string}
       */
      root: process.env.NODE_ENV === 'development' ? 'dev' : 'prd',

      /**
       * 模版文件路径
       * 相对于 `dist.root` 的相对地址
       * 若不区分布局文件和网页文件，请直接传入字符串
       * @type {(string|object)}
       */
      templates: {
        layout: 'templates/layout',
        pages: 'templates/pages'
      },

      /**
       * JavaScript 输出目录
       * @type {string}
       */
      js: 'js',

      /**
       * CSS 输出目录
       * @type {string}
       */
      css: 'css'
    },

    /**
     * 页面初始化 mock 数据文件存放目录
     * @type {string}
     */
    mockPages: 'mock/pages',

    /**
     * dllPlugin 编译输出物临时存放目录
     * @type {string}
     */
    tmpDll: '.tmp/dll',

    /**
     * 打包入口文件
     * @type {(string|object|function)}
     * @example
     * // string
     * entries: './src/entries/index.js'
     * @example
     * // object
     * entries: {
     *   index: './src/entries/index.js',
     *   abc: './src/entries/abc.less'
     * }
     * @example
     * // function
     * entries: () => {}
     */
    entries: () => {
      const entryFileName = 'entry.js';
      const entryPath = 'src/pages';
      const entryPattern = `**/${entryFileName}`;
      const cwd = path.resolve(context, entryPath);
      const config = {};
      packingGlob(entryPattern, { cwd }).forEach((page) => {
        const key = page.replace(`/${entryFileName}`, '');
        config[key] = path.join(cwd, page);
      });
      return config;
    }
  },

  /** 模版配置 */
  template: {
    /**
     * 是否启用 packing template
     * @type {bool}
     */
    enabled: true,

    /**
     * packing template 选项
     * @type {object}
     */
    options: {
      /**
       * 模版引擎类型
       * 目前支持
       * - html
       * - pug
       * - ejs
       * - handlebars
       * - smarty
       * - velocity
       * @type {string}
       */
      engine: 'pug',

      /**
       * 模版文件扩展名
       * @type {string}
       */
      extension: '.pug',

      /**
       * 是否根据 `entry pointer` 自动生成网页文件
       * 如需兼容 packing@<3.0.0 的工程，该值设置为 false
       * @type {bool}
       */
      autoGeneration: true,

      /**
       * 是否往模版中注入 assets
       * @type {bool}
       */
      inject: true,

      /**
       * JavaScript Chunk 注入的位置
       * - 'head': 在</head>前注入
       * - 'body': 在</body>前注入
       * @type {'head'|'body'}
       */
      scriptInjectPosition: 'body',

      /**
       * 是否往模版中注入 PWA manifest.json
       * @type {bool}
       */
      injectManifest: false,

      /**
       * `manifest.json` 输出位置
       * @type {string}
       */
      manifest: 'manifest.json',

      /**
       * 母模版位置
       * @type {string}
       */
      master: 'src/templates/pages/default.pug',

      /**
       * 输出网页使用的字符编码
       * @type {string}
       */
      charset: 'UTF-8',

      /**
       * 输出网页使用的标题
       * @type {string}
       */
      title: '',

      /**
       * 输出网页使用的 favicon 图标
       * - false: 不使用 favicon 图标
       * - 非空字符串: favicon 图标的位置
       * @type {(bool|string)}
       */
      favicon: false,

      /**
       * 输出网页使用的关键字
       * @type {(bool|string)}
       */
      keywords: false,

      /**
       * 输出网页使用的描述
       * @type {(bool|string)}
       */
      description: false,

      /**
       * 网页文件中需要在编译时替换为 _hash 的标签属性列表
       * 格式为 tag:attribute
       * 如果想对所有标签的某个属性替换，请使用 * 代替 tag
       * 如所有标签的 src 属性都需要替换，则使用 *:src
       * @example ['*:src', 'link:href']
       * @type {array}
       */
      attrs: ['img:src', 'link:href'],

      /**
       * 模版中命中的静态文件编译输出的文件名
       * @type {string}
       */
      path: '[path][name]_[hash:8].[ext]'
    }

  },

  /** HRM 配置 */
  hot: {
    /**
     * 是否启用热模块替换
     * @type {bool}
     */
    enabled: true,

    /**
     * HRM 选项
     * @type {object}
     * @see {@link https://github.com/webpack-contrib/webpack-hot-middleware|webpack-hot-middleware}
     */
    options: {}
  },

  /** 长效缓存配置 */
  longTermCaching: {
    /**
     * 是否启用编译时文件 hash 重命名
     * @type {bool}
     */
    enabled: true,

    /**
     * 缓存选项
     * @type {object}
     */
    options: {
      /**
       * 文件名与 hash 连接使用的字符串
       * @type {string}
       */
      delimiter: '_',

      /**
       * hash 长度
       * @type {number}
       */
      fileHashLength: 8
    }
  },

  /** 压缩代码配置 */
  minimize: {
    /**
     * 是否压缩代码
     * @type {bool}
     */
    enabled: true,

    /**
     * uglifyjs plugin 配置
     * @type {object}
     */
    options: {
      // sourceMap: true,
      uglifyOptions: {
        output: {
          // beautify: true,
          // 删除注释代码
          comments: false
        }
      }
    }
  },

  /**
   * `css-loader` 配置项
   * @type {object}
   * @see {@link https://github.com/webpack-contrib/css-loader|css-loader}
   */
  cssLoader: {
    /**
     * 在 css loader 之前应用的 loader 数量
     * @type {number}
     */
    importLoaders: 2,
    /**
     * 是否启用 `CSS Modules`
     * @type {bool}
     */
    modules: false,

    /**
     * 自定义css-modules类标识命名规则
     * @type {string}
     */
    localIdentName: '[path][name]__[local]--[hash:base64:5]'
  },

  /** stylelint 配置 */
  stylelint: {
    /**
     * 是否启用 `stylelint`
     * @type {bool}
     */
    enabled: true,

    /**
     * `stylelint` 配置项
     * @type {object}
     * @see {@link https://stylelint.io/user-guide/node-api/#options|stylelint options}
     */
    options: {
      files: ['**/*.css', '**/*.less', '**/*.s?(a|c)ss']
    }
  },

  /** eslint 配置 */
  eslint: {
    /**
     * 是否启用 `eslint`
     * @type {bool}
     */
    enabled: true,

    /**
     * `eslint` 配置项
     * @type {object}
     * @see {@link https://github.com/webpack-contrib/eslint-loader|eslint options}
     */
    options: {
    }
  },

  /**
   * runtimeChunk 配置
   * @see https://webpack.js.org/plugins/split-chunks-plugin/
   */
  runtimeChunk: {
    /**
     * 是否启用 runtimeChunk
     * @type {bool}
     */
    enabled: false,

    /**
     * runtimeChunk 输出的文件名
     * @type {string}
     */
    name: 'runtime'
  },

  /**
   * commonChunks 配置
   * 可以配置多个 common 包，配置的包名称会转换为正则表达式
   * 该配置分别在以下过程中被调用：
   * - 在 `packing serve` 任务中被 `DllPlugin` 调用
   * - 在 `packing build` 任务中被 `SplitChunkPlugin` 调用
   * 注意，如果配置了commonChunks，所有网页模版需要引用公共包文件
   * 否则会报错
   * <script src="/vendor.js"></script>
   * @type {object}
   * @see https://webpack.js.org/plugins/split-chunks-plugin/
   */
  commonChunks: {
    // vendor: [
    //   'react',
    //   'react-dom'
    //   'packing-ajax'
    // ]
  },

  /**
   * webpack-visualizer-plugin 配置
   * @see https://github.com/chrisbateman/webpack-visualizer
   */
  visualizer: {
    /**
     * 是否启用 webpack-visualizer-plugin
     * @type {bool}
     */
    enabled: true,

    /**
     * `visualizer` 配置项
     * @type {object}
     */
    options: {
      /**
       * 是否在浏览器中打开 visualizer 报表网页
       * 和 `packing build -o` 效果一样
       * @type {object}
       */
      open: process.env.npm_lifecycle_event === 'build'
    }
  },

  /** graphql 配置 */
  graphql: {
    /**
     * 是否使用 `GraphQL-mock-server`
     * @type {bool}
     */
    enabled: false,

    options: {
      /**
       * GraphQL 地址
       * @type {string}
       */
      graphqlEndpoint: '/graphql',

      /**
       * GraphiQL 地址
       * @type {string}
       */
      graphiqlEndpoint: '/graphiql'
    }
  },

  /**
   * 静态资源类型
   * @type {array}
   */
  assetExtensions: [
    'jpg',
    'jpeg',
    'png',
    'gif',
    'mp3',
    'ttf',
    'woff',
    'woff2',
    'eot',
    'svg'
  ],

  /**
   * URL转发路由规则配置
   * `require!` 表示使用本地 mock 文件
   * @type {object}
   */
  rewriteRules: {
    /** API转发 */
    '^/api/(.*)': 'require!/mock/api/$1.js'
  }
};
```


## 迁移到 packing 3
### 修改点
* 升级依赖包
  * packing@latest
  * packing-template-pug@^2.0.4
  * packing-urlrewrite@^0.1.8
* 用 `file-loader` 代替 `url-loader` 加载静态文件
* 使用 `dotenv` 加载 `process.env` 环境变量
* 将 `packing-template` 合并到 `packing` 工程，方便开发调试
* 调整 `packing.path` 结构
    - `path.dll` 更名为 `path.tmpDll`
    - 删除 `assetsDist` `templatesDist` `templatesPagesDist`
    - 删除 `packing.path.assets` 配置，如果希望直接 `import` `path.assets` 下的文件请使用 `webpack.resolve.alias` 设置
      ```js
      webpackConfig.resolve.alias = {
        assets: 'assets'
      };
      ```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.5.7](https://github.com/packingjs/generator-packing/compare/v3.5.6...v3.5.7) (2020-10-20)

### [3.5.6](https://github.com/packingjs/generator-packing/compare/v3.5.5...v3.5.6) (2020-10-19)


### Bug Fixes

* 删除 build 命令中的 -o 参数 ([f0bb6f7](https://github.com/packingjs/generator-packing/commit/f0bb6f7732a5857e3f1fee0ccb6f83f0dcbc601f))

### [3.5.5](https://github.com/packingjs/generator-packing/compare/v3.5.4...v3.5.5) (2020-03-24)

### [3.5.4](https://github.com/packingjs/generator-packing/compare/v3.5.3...v3.5.4) (2020-03-12)

### [3.5.3](https://github.com/packingjs/generator-packing/compare/v3.5.2...v3.5.3) (2020-02-08)

### [3.5.2](https://github.com/packingjs/generator-packing/compare/v3.5.1...v3.5.2) (2019-11-15)


### Bug Fixes

* fix unlink error ([a1b722a](https://github.com/packingjs/generator-packing/commit/a1b722a3f5f3026ad612cb8f87a748124bd52e0e))

### [3.5.1](https://github.com/packingjs/generator-packing/compare/v3.5.0...v3.5.1) (2019-10-31)


### Bug Fixes

* 遗漏了 lint-staged 节点 ([a71fce6](https://github.com/packingjs/generator-packing/commit/a71fce63e335dd6aae8a40ade3847da38f529162))

## [3.5.0](https://github.com/packingjs/generator-packing/compare/v3.4.0...v3.5.0) (2019-10-31)


### Features

* precommit 时使用 lint-staged 对修改的文件做检查 ([268c2a7](https://github.com/packingjs/generator-packing/commit/268c2a70a366faea80f3eb545023700aa2ef29e9))

## [3.4.0](https://github.com/packingjs/generator-packing/compare/v3.3.1...v3.4.0) (2019-10-30)


### Features

* 支持 typescript ([f26fe4a](https://github.com/packingjs/generator-packing/commit/f26fe4ae3cdffc285137a02445adf35f7d4b7c06))
* 支持用 stylelint 检查 SASS 文件 ([afb4142](https://github.com/packingjs/generator-packing/commit/afb41429161ab9c202de232a7401e9aee5aa6c47))
* 默认支持 IE10 CSS ([ed13110](https://github.com/packingjs/generator-packing/commit/ed131107b13631b019dc773f5d20dacf634f037e))

<a name="3.3.1"></a>
## [3.3.1](https://github.com/packingjs/generator-packing/compare/v3.3.0...v3.3.1) (2019-08-29)



<a name="3.3.0"></a>
# [3.3.0](https://github.com/packingjs/generator-packing/compare/v3.2.4...v3.3.0) (2018-12-01)


### Features

* 支持 babel 7 ([06fb3d5](https://github.com/packingjs/generator-packing/commit/06fb3d5))



<a name="3.2.4"></a>
## [3.2.4](https://github.com/packingjs/generator-packing/compare/v3.2.3...v3.2.4) (2018-07-31)


### Bug Fixes

* 修复编译失败时不终止 jenkins 的问题，导致错误代码发布到线上环境 ([7b1cd3c](https://github.com/packingjs/generator-packing/commit/7b1cd3c))



<a name="3.2.3"></a>
## [3.2.3](https://github.com/packingjs/generator-packing/compare/v3.2.2...v3.2.3) (2018-06-27)



<a name="3.2.2"></a>
## [3.2.2](https://github.com/packingjs/generator-packing/compare/v3.2.1...v3.2.2) (2018-05-16)



<a name="3.2.1"></a>
## [3.2.1](https://github.com/packingjs/generator-packing/compare/v3.2.0...v3.2.1) (2018-05-16)


### Bug Fixes

* async/await 语法报错 ([862b67b](https://github.com/packingjs/generator-packing/commit/862b67b))



<a name="3.2.0"></a>
# [3.2.0](https://github.com/packingjs/generator-packing/compare/v3.1.2...v3.2.0) (2018-05-16)


### Features

* 增加编译平台选项 ([a520a3f](https://github.com/packingjs/generator-packing/commit/a520a3f))



<a name="3.1.2"></a>
## [3.1.2](https://github.com/packingjs/generator-packing/compare/v3.1.1...v3.1.2) (2018-05-14)



<a name="3.1.1"></a>
## [3.1.1](https://github.com/packingjs/generator-packing/compare/v3.1.0...v3.1.1) (2018-05-10)



<a name="3.1.0"></a>
# [3.1.0](https://github.com/packingjs/generator-packing/compare/v3.0.7...v3.1.0) (2018-05-10)


### Features

* 删除对 artTemplate 的支持 ([36e8462](https://github.com/packingjs/generator-packing/commit/36e8462))



<a name="3.0.7"></a>
## [3.0.7](https://github.com/packingjs/generator-packing/compare/v3.0.6...v3.0.7) (2018-05-04)


### Bug Fixes

* 修复 sass 文件语法错误 ([8bedf04](https://github.com/packingjs/generator-packing/commit/8bedf04))



<a name="3.0.6"></a>
## [3.0.6](https://github.com/packingjs/generator-packing/compare/v3.0.5...v3.0.6) (2018-05-03)



<a name="3.0.5"></a>
## [3.0.5](https://github.com/packingjs/generator-packing/compare/v3.0.4...v3.0.5) (2018-05-03)



<a name="3.0.4"></a>
## [3.0.4](https://github.com/packingjs/generator-packing/compare/v3.0.3...v3.0.4) (2018-05-03)



<a name="3.0.3"></a>
## [3.0.3](https://github.com/packingjs/generator-packing/compare/v3.0.2...v3.0.3) (2018-05-03)



<a name="3.0.2"></a>
## [3.0.2](https://github.com/packingjs/generator-packing/compare/v3.0.1...v3.0.2) (2018-04-26)



<a name="3.0.1"></a>
## [3.0.1](https://github.com/packingjs/generator-packing/compare/v2.9.14...v3.0.1) (2018-04-20)



<a name="2.9.14"></a>
## [2.9.14](https://github.com/packingjs/generator-packing/compare/v2.9.13...v2.9.14) (2018-02-12)



<a name="2.9.13"></a>
## [2.9.13](https://github.com/packingjs/generator-packing/compare/v2.9.12...v2.9.13) (2018-02-01)



<a name="2.9.12"></a>
## [2.9.12](https://github.com/packingjs/generator-packing/compare/v2.9.11...v2.9.12) (2018-01-26)



<a name="2.9.11"></a>
## [2.9.11](https://github.com/packingjs/generator-packing/compare/v2.9.10...v2.9.11) (2018-01-26)



<a name="2.9.10"></a>
## [2.9.10](https://github.com/packingjs/generator-packing/compare/v2.9.9...v2.9.10) (2018-01-25)


### Bug Fixes

* 修复cdnroot路径，在末端加上反斜杠 ([e02f54f](https://github.com/packingjs/generator-packing/commit/e02f54f))



<a name="2.9.9"></a>
## [2.9.9](https://github.com/packingjs/generator-packing/compare/v2.9.3...v2.9.9) (2018-01-17)


### Bug Fixes

* bablerc 中增加stage-0 ([f072602](https://github.com/packingjs/generator-packing/commit/f072602))



<a name="2.9.3"></a>

generators/app/templates/config/packing.js
/**
 * 这个文件可以修改packing配置文件的默认设置
 * 默认配置请看 `node_modules/packing/config/packing.js`
 *
 * @param object packing 默认配置对象
 */<% var templateExtension;
 switch (props.template) {
   case 'ejs':
     templateExtension = 'ejs';
     break;
   case 'handlebars':
     templateExtension = 'hbs';
     break;
   case 'pug':
     templateExtension = 'pug';
     break;
   case 'smarty':
     templateExtension = 'tpl';
     break;
   case 'velocity':
     templateExtension = 'vm';
     break;
   default:
     templateExtension = 'html';
     break;
 } %>

export default (packing) => {
  const p = packing;<% if (props.template !== 'pug') { %>

  // 模版引擎类型，目前支持的类型有[html,pug,ejs,handlebars,smarty,velocity]
  p.template.options.engine = '<%= props.template%>';

  // 模版文件扩展名
  p.template.options.extension = '.<%= templateExtension%>';

  // 通用模版位置
  p.template.options.master = 'src/templates/pages/default.<%= templateExtension%>';<% } %>

  // 网站自定义配置
  p.rewriteRules = {
    // 网站URL与模版的对应路由关系
    '^/$': '/index',
    // API转发
    '^/api/(.*)': 'require!/mock/api/$1.js'
  };

  // 是否启用打包分析工具
  p.bundleAnalyzer.enabled = false;

  return p;
};

/** 所有配置项及默认值 */
// {
//   /**
//    * 本地访问的域名
//    * 如果需要使用 `qunar.com` 的 cookie，需要改成类似 `my.qunar.com` 这种
//    * @type {string}
//    */
//   localhost: 'localhost',
//
//   /**
//    * webserver 端口
//    */
//   port: {
//     /**
//      * 开发环境 webserver 端口
//      * @type {number}
//      */
//     dev: 8081,
//
//     /**
//      * 预览编译结果时 webserver 端口
//      * @type {number}
//      */
//     dist: 8080
//   },
//
//   /**
//    * 文件路径配置
//    * 所有目录都使用相对于项目根目录的相对目录格式
//    */
//   path: {
//     /** 源文件相关路径 */
//     src: {
//       /**
//        * 源文件根目录
//        * @type {string}
//        */
//       root: 'src',
//
//       /**
//        * 模版文件路径
//        * 相对于 `src.root` 的相对地址
//        * 若不区分布局文件和网页文件，请直接传入字符串
//        * @type {(string|object)}
//        */
//       templates: {
//         layout: 'templates/layout',
//         pages: 'templates/pages'
//       }
//     },
//
//     /** 编译输出文件相关路径 */
//     dist: {
//       /**
//        * webpack 编译产物输出目录
//        * 即 `webpack.config.output.path` 参数
//        * portal dev 发布时要求输出到 `dev` 目录
//        * qdr dev 发布时要求输出到 `prd` 目录
//        * @type {string}
//        */
//       root: process.env.NODE_ENV === 'development' ? 'dev' : 'prd',
//
//       /**
//        * 模版文件路径
//        * 相对于 `dist.root` 的相对地址
//        * 若不区分布局文件和网页文件，请直接传入字符串
//        * @type {(string|object)}
//        */
//       templates: {
//         layout: 'templates/layout',
//         pages: 'templates/pages'
//       },
//
//       /**
//        * JavaScript 输出目录
//        * @type {string}
//        */
//       js: 'js',
//
//       /**
//        * CSS 输出目录
//        * @type {string}
//        */
//       css: 'css'
//     },
//
//     /**
//      * 页面初始化 mock 数据文件存放目录
//      * @type {string}
//      */
//     mockPages: 'mock/pages',
//
//     /**
//      * dllPlugin 编译输出物临时存放目录
//      * @type {string}
//      */
//     tmpDll: '.tmp/dll',
//
//     /**
//      * 打包入口文件
//      * @type {(string|object|function)}
//      * @example
//      * // string
//      * entries: './src/entries/index.js'
//      * @example
//      * // object
//      * entries: {
//      *   index: './src/entries/index.js',
//      *   abc: './src/entries/abc.less'
//      * }
//      * @example
//      * // function
//      * entries: () => {}
//      */
//     entries: () => {
//       const entryFileName = 'entry.js';
//       const entryPath = 'src/pages';
//       const entryPattern = `**/${entryFileName}`;
//       const cwd = path.resolve(context, entryPath);
//       const config = {};
//       packingGlob(entryPattern, { cwd }).forEach((page) => {
//         const key = page.replace(`/${entryFileName}`, '');
//         config[key] = path.join(cwd, page);
//       });
//       return config;
//     }
//   },
//
//   /** 模版配置 */
//   template: {
//     /**
//      * 是否启用 packing template
//      * @type {bool}
//      */
//     enabled: true,
//
//     /**
//      * packing template 选项
//      * @type {object}
//      */
//     options: {
//       /**
//        * 模版引擎类型
//        * 目前支持
//        * - html
//        * - pug
//        * - ejs
//        * - handlebars
//        * - smarty
//        * - velocity
//        * @type {string}
//        */
//       engine: 'pug',
//
//       /**
//        * 模版文件扩展名
//        * @type {string}
//        */
//       extension: '.pug',
//
//       /**
//        * 是否根据 `entry pointer` 自动生成网页文件
//        * 如需兼容 packing@<3.0.0 的工程，该值设置为 false
//        * @type {bool}
//        */
//       autoGeneration: true,
//
//       /**
//        * 是否往模版中注入 assets
//        * @type {bool}
//        */
//       inject: true,
//
//       /**
//        * JavaScript Chunk 注入的位置
//        * - 'head': 在</head>前注入
//        * - 'body': 在</body>前注入
//        * @type {'head'|'body'}
//        */
//       scriptInjectPosition: 'body',
//
//       /**
//        * 是否往模版中注入 PWA manifest.json
//        * @type {bool}
//        */
//       injectManifest: false,
//
//       /**
//        * `manifest.json` 输出位置
//        * @type {string}
//        */
//       manifest: 'manifest.json',
//
//       /**
//        * 母模版位置
//        * @type {string}
//        */
//       master: 'src/templates/pages/default.pug',
//
//       /**
//        * 输出网页使用的字符编码
//        * @type {string}
//        */
//       charset: 'UTF-8',
//
//       /**
//        * 输出网页使用的标题
//        * @type {string}
//        */
//       title: '',
//
//       /**
//        * 输出网页使用的 favicon 图标
//        * - false: 不使用 favicon 图标
//        * - 非空字符串: favicon 图标的位置
//        * @type {(bool|string)}
//        */
//       favicon: false,
//
//       /**
//        * 输出网页使用的关键字
//        * @type {(bool|string)}
//        */
//       keywords: false,
//
//       /**
//        * 输出网页使用的描述
//        * @type {(bool|string)}
//        */
//       description: false,
//
//       /**
//        * 网页文件中需要在编译时替换为 _hash 的标签属性列表
//        * 格式为 tag:attribute
//        * 如果想对所有标签的某个属性替换，请使用 * 代替 tag
//        * 如所有标签的 src 属性都需要替换，则使用 *:src
//        * @example ['*:src', 'link:href']
//        * @type {array}
//        */
//       attrs: ['img:src', 'link:href'],
//
//       /**
//        * 模版中命中的静态文件编译输出的文件名
//        * @type {string}
//        */
//       path: '[path][name]_[hash:8].[ext]'
//     }
//
//   },
//
//   /** HRM 配置 */
//   hot: {
//     /**
//      * 是否启用热模块替换
//      * @type {bool}
//      */
//     enabled: true,
//
//     /**
//      * HRM 选项
//      * @type {object}
//      * @see {@link https://github.com/webpack-contrib/webpack-hot-middleware|webpack-hot-middleware}
//      */
//     options: {}
//   },
//
//   /** 长效缓存配置 */
//   longTermCaching: {
//     /**
//      * 是否启用编译时文件 hash 重命名
//      * @type {bool}
//      */
//     enabled: true,
//
//     /**
//      * 缓存选项
//      * @type {object}
//      */
//     options: {
//       /**
//        * 文件名与 hash 连接使用的字符串
//        * @type {string}
//        */
//       delimiter: '_',
//
//       /**
//        * hash 长度
//        * @type {number}
//        */
//       fileHashLength: 8
//     }
//   },
//
//   /** 压缩代码配置 */
//   minimize: {
//     /**
//      * 是否压缩代码
//      * @type {bool}
//      */
//     enabled: true,
//
//     /**
//      * uglifyjs plugin 配置
//      * @type {object}
//      */
//     options: {
//       // sourceMap: true,
//       uglifyOptions: {
//         output: {
//           // beautify: true,
//           // 删除注释代码
//           comments: false
//         }
//       }
//     }
//   },
//
//   /**
//    * `css-loader` 配置项
//    * @type {object}
//    * @see {@link https://github.com/webpack-contrib/css-loader|css-loader}
//    */
//   cssLoader: {
//     /**
//      * 在 css loader 之前应用的 loader 数量
//      * @type {number}
//      */
//     importLoaders: 2,
//     /**
//      * 是否启用 `CSS Modules`
//      * @type {bool}
//      */
//     modules: false,
//
//     /**
//      * 自定义css-modules类标识命名规则
//      * @type {string}
//      */
//     localIdentName: '[path][name]__[local]--[hash:base64:5]'
//   },
//
//   /** eslint 配置 */
//   eslint: {
//     /**
//      * 是否启用 `eslint`
//      * @type {bool}
//      */
//     enabled: true,
//
//     /**
//      * `eslint` 配置项
//      * @type {object}
//      * @see {@link https://github.com/webpack-contrib/eslint-loader|eslint options}
//      */
//     options: {
//     }
//   },
//
//   /**
//    * runtimeChunk 配置
//    * @see https://webpack.js.org/plugins/split-chunks-plugin/
//    */
//   runtimeChunk: {
//     /**
//      * 是否启用 runtimeChunk
//      * @type {bool}
//      */
//     enabled: false,
//
//     /**
//      * runtimeChunk 输出的文件名
//      * @type {string}
//      */
//     name: 'runtime'
//   },
//
//   /**
//    * commonChunks 配置
//    * 可以配置多个 common 包，配置的包名称会转换为正则表达式
//    * 该配置分别在以下过程中被调用：
//    * - 在 `packing serve` 任务中被 `DllPlugin` 调用
//    * - 在 `packing build` 任务中被 `SplitChunkPlugin` 调用
//    * 注意，如果配置了commonChunks，所有网页模版需要引用公共包文件
//    * 否则会报错
//    * <script src="/vendor.js"></script>
//    * @type {object}
//    * @see https://webpack.js.org/plugins/split-chunks-plugin/
//    */
//   commonChunks: {
//     // vendor: [
//     //   'react',
//     //   'react-dom'
//     //   'packing-ajax'
//     // ]
//   },
//
//   /**
//    * webpack-visualizer-plugin 配置
//    * @see https://github.com/chrisbateman/webpack-visualizer
//    */
//   visualizer: {
//     /**
//      * 是否启用 webpack-visualizer-plugin
//      * @type {bool}
//      */
//     enabled: true,
//
//     /**
//      * `visualizer` 配置项
//      * @type {object}
//      */
//     options: {
//       /**
//        * 是否在浏览器中打开 visualizer 报表网页
//        * 和 `packing build -o` 效果一样
//        * @type {object}
//        */
//       open: process.env.npm_lifecycle_event === 'build'
//     }
//   },
//
//   /** graphql 配置 */
//   graphql: {
//     /**
//      * 是否使用 `GraphQL-mock-server`
//      * @type {bool}
//      */
//     enabled: false,
//
//     options: {
//       /**
//        * GraphQL 地址
//        * @type {string}
//        */
//       graphqlEndpoint: '/graphql',
//
//       /**
//        * GraphiQL 地址
//        * @type {string}
//        */
//       graphiqlEndpoint: '/graphiql'
//     }
//   },
//
//   /**
//    * 静态资源类型
//    * @type {array}
//    */
//   assetExtensions: [
//     'jpg',
//     'jpeg',
//     'png',
//     'gif',
//     'mp3',
//     'ttf',
//     'woff',
//     'woff2',
//     'eot',
//     'svg'
//   ],
//
//   /**
//    * URL转发路由规则配置
//    * `require!` 表示使用本地 mock 文件
//    * @type {object}
//    */
//   rewriteRules: {
//     /** API转发 */
//     '^/api/(.*)': 'require!/mock/api/$1.js'
//   }
//
// /**
//  * packing-serve使用的webpackDevMiddleware配置项
//  * @type {object}
//  */
// devMwOptions: {
//   /**
//    * 是否将中间产物写入文件
//    * @type {bool|function}
//    */
//   writeToDisk: false
// },
//
// /**
//  * packing-serve是否支持部分页面编译
//  * @type {object}
//  */
// partialCompile: {
//   /**
//    * 是否启用部分页面编译
//    * @type {bool}
//    */
//   enabled: false,
//
//   /**
//    * 默认编译的页面
//    * @type {array}
//    */
//   whitelist: []
// }
